### PR TITLE
PoC: Jenkinsfile+dobi for CI

### PR DESCRIPTION
**Pull request by dnephin**
The goal of this PR is to show the use of Jenkinsfile for configuring our CI builds. Jenkinsfile gives us a way of programmatically configuring build pipelines from the source repo.

In addition to the Jenkinsfile, the following supporting changes were made:
### Move validation scripts to hack/validate

The `hack/make` framework of creating bundle directories and requiring execution of scripts to be run from a driver (`hack/make.sh`) is "ok" for build tasks that produce artifacts, but has little to no benefit for these validation/check scripts. They were changed to source all their dependencies directly, so they no longer require the `hack/make.sh` driver. They were moved out `hack/make` to `hack/validate` so that it's clear they can be run directly without the driver.

They also run as a single process now, so instead of every one of them having to perform a `git fetch`, `hack/validate/all` does a single `git fetch`, reducing the overall runtime of validation.
### Independent images for different (groups of) build tasks

Instead of a single monolithic `Dockerfile` for every build task, we can split them into separate Dockerfiles for groups of related tasks. This comes with a few benefits:
- running many task is significantly faster. For example, running the `validate/all` task in a container, after a fresh checkout (or after a dependency in the image changes) only takes ~5-10s. Currently it can take easily 10+ minutes to build the monolithic image. Running tests or building the linux binary will also be faster, because we don't need to install all the dependencies required for other tasks (ex  #26389). This is not just a one-time improvement. Every time we make a dependency change to an image we invalidate everything after it. By splitting the images, a change to one doesn't force us to rebuild the image for all the other tasks. We end up with much better cache hit rates.
- maintaining dockerfiles for different platforms becomes easier. We don't need to install dependencies for every task. We only need to install dependencies for building the binary on that platform.

This PR adds two new images (from `Dockerfile.validate`, and `Dockerfile.build`),  #26389 has one for `cross`, and there is already a separate image for man page generation. In the near future we can have one for generating the swagger types (#27117). We might also want images for `build-rpm` and `build-dep`.
### dobi for running build tasks

[dobi](http://dnephin.github.io/dobi/) is a tool I've been working on for "build automation". It is designed based on my experienced writing Makefiles both here and at previous employer, as well as a lot of feedback from Compose users who are attempting to use Compose as a Makefile replacement (which it is not designed to be). dobi is aimed at being a make-like tool that exposes docker resources from a declarative config file. It would eventually be a replacement for the Makefile. 

As you can see from the initial `dobi.yaml` in this PR, you define each resource (mount, image, job) and you can execute them as you would a make target with `dobi <resource name>`. There are also "actions" for each task, so to remove the build image you could run `dobi builder:rm` (see the [task reference](http://dnephin.github.io/dobi/tasks.html)).

dobi comes with a long list of improvements over a Makefile:
- significantly better caching (http://dnephin.github.io/dobi/tasks.html#build-default)
- more intuitive syntax (I'm probably biased, but it does at least seem to be significantly less verbose than the Makefile version)

If you're interested in trying it out , you can run `dobi list` to get a list of all the resources that have descriptions.

### CODE REVIEW

**Comment by dnephin** [comment]
There's a lot going on in this PR. If there are pieces that you think are less controversial I can split them out into a separate PR to get some of it merged faster.

**Comment by vdemeester** [comment]
One thing I'm not *that found of*, is that now I need `dobi` on my system. It adds one more dependencies to be able to build `docker`. `docker` and `make` are in system packages (`apt-get`, `dnf`, …), `dobi` is not really… so it does add a bigger barrier to being able to build docker itself.

**Comment by dnephin** [comment]
Yes, it's true that it's a new dependency. I can create an `rpm` and `deb` for `dobi` to make it easier to install on those systems.

**Comment by mikedougherty** [comment]
the general thrust of the change is fine with me; the only thing i think my opinion should count for is the Jenkinsfile which LGTM :)

one future improvement i might suggest is that "failing" output for a validate task (e.g.: a golint warning) should go to a file on disk that can be archived by the CI system. this can then be interpreted over time in various ways if desired.

**Comment by bfirsh** [comment]
Would it be possible to run dobi inside a container with a bind mounted Docker socket?

**Comment by bfirsh** [comment]
That sounds like a better solution, I think. I don't think running a small bash script to `docker run` dobi with the correct volumes is worse – bash installed as standard on all developer machines. Then, the only dependency to build Docker is Docker.

**Comment by dnephin** [comment]
I guess if anyone has `dobi` installed they can still run things directly. The script would only be necessary if you want to run it in a container. I'm good with that.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,28 @@
+
+
+def runTask(String label, String task) {
+  { ->
+    wrappedNode(label: label) {
+      deleteDir()
+      checkout scm
+      withEnv(["APT_MIRROR=cdn-fastly.deb.debian.org"]) {
```
Comment (glensc) [inline Jenkinsfile]: afaik `Dockerfile` used `ARG` not `ENV` for `APT_MIRROR`

**Comment by mlaventure** [comment]
ping @dnephin where is this at?

I'm also for hiding dobi within a container behind a script if that matters :)

**Comment by vdemeester** [comment]
@mlaventure @dnephin I think we should split up this PR in two : one that add the Jenkinsfile, the other that uses `dobi`. I'll experiment a bit (with @dnephin 👼)

**Comment by mlaventure** [comment]
@vdemeester SGTM. I'd like to unblock #26389 though

**Comment by dnephin** [comment]
I opened #28952 to replace this

**Pull request closed without merging**